Add REFRESH options for MVs -- parsing, purification, planning

ggevay opened the pull request:
This is the parsing, purification, planning of REFRESH options on materialized views. [This is the full PR, which includes the Compute part.](https://github.com/MaterializeInc/materialize/pull/23819)

[Design document.](https://github.com/ggevay/materialize/blob/refresh-every-design-doc/doc/developer/design/20231027_refresh_every_mv.md#solution-proposal)

### Motivation

<!--
Which of the following best describes the motivation behind this PR?

  * This PR fixes a recognized bug.

    [Ensure issue is linked somewhere.]

  * This PR adds a known-desirable feature.

    [Ensure issue is linked somewhere.]

  * This PR fixes a previously unreported bug.

    [Describe the bug in detail, as if you were filing a bug report.]

  * This PR adds a feature that has not yet been specified.

    [Write a brief specification for the feature, including justification
     for its inclusion in Materialize, as if you were writing the original
     feature specification.]

   * This PR refactors existing code.

    [Describe what was wrong with the existing code, if it is not obvious.]
-->

### Tips for reviewer

<!--
Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.
-->

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

benesch reviewed (commented):
The code here looks really quite solid. I left you a bunch of comments around parser structure and error message style that should be straightforward to fix and don't fundamentally alter the nature of the PR. 

There are two harder comments:

  1. I'd love to try to simplify the purification plumbing—specifically, reusing the existing `purify_statement` method to avoid introducing materialized view-specific public API to the `sql` crate. ISTM this might be straightforward by not caring about non-EpochMillis timelines. Holler if that's not the case.
  2. Seeing Joe's comment about DST with intervals and your comment about monthly intervals makes me nervous about supporting `REFRESH EVERY`. I think we should have a quick discussion on the Slack, wherever makes sense, about whether we should change gears and try `REFRESH ON SCHEDULE` instead.

benesch commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -2086,9 +2090,125 @@ pub fn plan_create_materialized_view(
+                    if interval.as_microseconds() <= 0 {
+                        sql_bail!("REFRESH interval must be positive; got: {}", interval);
+                    }
+                    if interval.as_microseconds() > Interval::new(0, 27, 0).as_microseconds() {
```
Oh, maybe the DST thing is not an issue because we say the schedule is always computed in UTC?

benesch replied:
Ok, so looking at this more deeply, I don't think we need to ban intervals longer than 30 days. An interval of length 90 days is totally fine as long as the user said `90 days` and not `3mo`. I think the condition we actually want to check here is that `interval.months == 0`. And then we can just call `Interval::duration()?` and we're good; as mentioned above I think it's fine to say all days have exactly 24 hours because we can just document that the schedule proceeds according to UTC regardless of whatever timezone you put in your `ALIGNED TO <timestamptz>` clause.

jkosh44 replied:
Would we need to care about leap seconds? The interval docs don't mention leap seconds anywhere but if we care then we may want to double check that they're ignored.

ggevay replied:
I think the problem doesn't occur with leap seconds, because those are handled by slight adjustments to the epoch millis timestamp rather than changing the mapping from human-readable times like 11:00 to epoch milliseconds, which is what a daylight saving time transition does.

ggevay replied:
Ok, I will
- check `interval.months == 0`
- call `Interval::duration()`
- document that all days are considered 24 hours, and call it out that this might mean a +-1 hour shift across DST boundaries.

(Eventually, I'll implement cron schedules too anyhow, because it's also needed for complex schedules like not refreshing on weekends. And at that point, if somebody doesn't like the +-1 hour shift of REFRESH EVERY, then they can use cron schedules instead.)

ggevay replied:
Done. (Except for the documentation part, which will be in a follow-up PR.)

benesch commented on src/repr/src/refresh_schedule.rs:
```diff
@@ -0,0 +1,35 @@
+
+#[derive(Clone, Debug, Serialize, Deserialize, Eq, PartialEq)]
+pub struct RefreshEvery {
+    pub interval: Interval,
```
```suggestion
    pub interval: Duration,
```

Recommend storing a `Duration` here, and converting from `Interval` to `Duration` in the SQL planner. That's what we do elsewhere. The `Interval` type is annoying as hell to use from Rust.

ggevay replied:
Ok, done: https://github.com/MaterializeInc/materialize/pull/23870/commits/67f93859f1cd8d85ec74c3343479230a4d313225

ggevay commented on src/sql/src/pure.rs:
```diff
@@ -1475,3 +1482,239 @@ async fn compile_proto(
+    // 5. Attend to `resolved_ids`: The purification might have
+    // - added references to `mz_timestamp`;
+    // - removed references to `mz_now`.
+    if introduced_mz_timestamp {
+        resolved_ids.0.insert(mz_timestamp_id);
+    }
+    // Even though we always remove `mz_now()` from the `with_options`, there might be `mz_now()`
+    // remaining in the main query expression of the MV, so let's visit the entire statement to look
+    // for `mz_now()` everywhere.
+    let mut visitor = ExprContainsTemporalVisitor::new();
+    visitor.visit_create_materialized_view_statement(cmvs);
+    if !visitor.contains_temporal {
+        resolved_ids.0.remove(&mz_now_id);
+    }
```
Unfortunately, they can't be wrong in either direction. I had a bug during development where they were wrong, and I was hitting the `in-memory state of the catalog does not match its on-disk state` check in Testdrive tests. This is because the `resolved_ids` are stored in the in-memory state, and if a wrong set is being stored, then the set won't match the new set that we get when we reconstruct the in-memory catalog state from disk.

jkosh44 replied:
What are these IDs used for?

benesch replied:
Ah, well, that seems plausible enough @ggevay. Let's leave the code as is then!

And @jkosh44 I was hoping you knew the answer to that question! I paged out the details long ago.

jkosh44 replied:
I was hoping that @ggevay knew since this is a new function introduced in this PR. I took a look and these are later plumbed into planning, which uses these to check if any function in `mz_unsafe` is called. It's later plumbed into sequencing which uses them to determine required RBAC privileges and is then saves them in the catalog. The catalog assumes these are IDs that are directly referenced by the SQL statement.

benesch replied:
Yeah, @ggevay was forced to handle them here, but the resolved IDs were already created by the coordinator _before_ Gabor inserted this function, and then handled by planning/the coordinator _after_ this function, so there are some un/underdocumented guarantees that planning/sequencing require.

jkosh44 replied:
I'm having trouble articulating exactly what the requirements are, but it's something like this.

When we receive a user DDL, we make the following transformation:
`user-ddl-sql -> (persisted-sql, resolved-ids)`. The `resolved_ids` input/output to this function is exactly those `resolved-ids`. The requirement is that when we re-parse and plan `persisted-sql` we should end up with the exact same set of `GlobalId`s that are in `resolved-ids`. That's (probably) why the catalog validations were failing in testdrive.

benesch replied:
Ah, that makes sense! We should write that down as a comment in the code.

jkosh44 replied:
https://github.com/MaterializeInc/materialize/pull/24016

ggevay commented on src/sql/src/pure.rs:
```diff
@@ -1475,3 +1482,239 @@ async fn compile_proto(
+            )));
+        }
+
+        // 2. If `REFRESH EVERY` doesn't have a `STARTING AT`, then add `STARTING AT mz_now()`.
```
fixed

ggevay commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -2086,9 +2090,125 @@ pub fn plan_create_materialized_view(
+        }
+
+        if on_commits_seen > 1 {
+            sql_bail!("REFRESH ON COMMIT can be given only once");
```
fixed

ggevay commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -2086,9 +2090,125 @@ pub fn plan_create_materialized_view(
+                    on_commits_seen += 1;
+                }
+                RefreshOptionValue::AtCreation => {
+                    unreachable!("REFRESH AT CREATION should have been purified away")
```
yep, fixed

ggevay commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -2086,9 +2090,125 @@ pub fn plan_create_materialized_view(
+                    };
+                    let hir = plan_expr(ecx, &time)?.cast_to(
+                        ecx,
+                        CastContext::Implicit,
```
ok, changed

ggevay commented on src/sql-parser/src/ast/defs/statement.rs:
```diff
@@ -3179,11 +3182,57 @@ impl<T: AstInfo> AstDisplay for WithOptionValue<T> {
             WithOptionValue::ConnectionKafkaBroker(broker) => {
                 f.write_node(broker);
             }
+            WithOptionValue::Refresh(RefreshOptionValue::OnCommit) => {
+                f.write_str("ON COMMIT");
+            }
+            WithOptionValue::Refresh(RefreshOptionValue::AtCreation) => {
+                f.write_str("AT CREATION");
+            }
+            WithOptionValue::Refresh(RefreshOptionValue::At(RefreshAtOptionValue { time })) => {
+                f.write_str("AT ");
+                f.write_node(time);
+            }
+            WithOptionValue::Refresh(RefreshOptionValue::Every(RefreshEveryOptionValue {
+                interval,
+                aligned_to,
+            })) => {
+                f.write_str("EVERY '");
+                f.write_str(interval);
+                f.write_str("'");
+                if let Some(aligned_to) = aligned_to {
+                    f.write_str(" ALIGNED TO ");
+                    f.write_node(aligned_to)
+                }
+            }
```
Done

ggevay commented on src/sql-parser/src/ast/defs/statement.rs:
```diff
@@ -3179,11 +3182,57 @@ impl<T: AstInfo> AstDisplay for WithOptionValue<T> {
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct RefreshEveryOptionValue<T: AstInfo> {
+    // The following is a String and not an IntervalValue, because that starts with the keyword
+    // INTERVAL, but that is not needed here, since the only thing that can come here is an
+    // interval, so no need to indicate this with an extra keyword.
```
One more question: I don't really understand how I should handle all those extra fields on `IntervalValue` besides the `IntervalValue`. It seems that `Duration::try_from_value` simply ignores them, if I'm reading the code correctly? Should I just call `Duration::try_from_value` in the planning?

ggevay replied:
@jkosh44, @benesch 
I made the refactoring: 38d6ffd1bdc2aa27e9631b0eccacc2942eb8a709 Could you please check it?

ggevay replied:
Ah, sorry, I forgot to change the comment!

ggevay replied:
(Fixed the comment now)

ggevay commented on src/sql-parser/src/parser.rs:
```diff
@@ -3218,18 +3218,102 @@ impl<'a> Parser<'a> {
+                            should NOT be present!"
+                        )
+                    )
+                }
```
See [above](https://github.com/MaterializeInc/materialize/pull/23870#discussion_r1432656569).

ggevay commented on src/adapter/src/coord/command_handler.rs:
```diff
@@ -683,6 +688,87 @@ impl Coordinator {
+                            Ok(tc) => tc,
+                            Err(e) => return ctx.retire(Err(e)),
+                        };
+                    let timeline = match timeline_context {
+                        TimelineContext::TimelineDependent(timeline) => timeline,
+                        TimelineContext::TimestampDependent
+                        | TimelineContext::TimestampIndependent => {
+                            // We default to EpochMilliseconds, similarly to `determine_timestamp_for`.
+                            // Note that we didn't accurately decide whether we are TimestampDependent
+                            // or TimestampIndependent, because for this we'd need to also check whether
+                            // `query.contains_temporal()`, similarly to how `peek_stage_validate` does.
+                            // However, this doesn't matter here, as we are just going to default to
+                            // EpochMilliseconds in both cases.
+                            Timeline::EpochMilliseconds
+                        }
+                    };
```
I've changed to `timeline_context.timeline().unwrap_or(Timeline::EpochMilliseconds)`.

> It may even be helpful to add a method to TimelineContext like timeline_or_default() that defaults to EpochMilliseconds. You could use that here and in determine_timestamp_for to prevent them from diverging.

The logic in `determine_timestamp_for` seems slightly different: It calls `get_timeline(timeline_context)`, which returns `None` for `TimestampIndependent`.

Should I still create `timeline_or_default()`, or it's not important if I would be the only caller?

ggevay commented on src/adapter/src/coord/command_handler.rs:
```diff
@@ -683,6 +688,87 @@ impl Coordinator {
+                        with_options: cmvs.with_options,
+                    });
+
+                // (Purifying CreateMaterializedView doesn't happen async, so no need to send
+                // `Message::PurifiedStatementReady` here.)
```
I don't think it's a slow process. It's just some AST traversal; no calling out to external systems. (I think that the other match case, which does it async, might communicate with Kafka or something.)

ggevay replied:
I've added a comment now.

ggevay replied:
I've extended the comment as discussed elsewhere.

ggevay commented on src/sql-parser/src/ast/defs/statement.rs:
```diff
@@ -3137,7 +3138,9 @@ impl<T: AstInfo> AstDisplay for WithOptionValue<T> {
+                WithOptionValue::Value(_)
+                | WithOptionValue::Sequence(_)
+                | WithOptionValue::Refresh(_) => {
                     // These are redact-aware.
```
Ok, I've added a redaction for the refresh interval. The other literals in `WithOptionValue::Refresh` are in `Expr`s, which will be redacted already, right?

ggevay replied:
(Sorry, despite the above question, I had to click "Resolve conversation" here due to a crazy GitHub glitch: for some reason, it keeps scrolling the page to this conversation every few seconds...)

ggevay replied:
Pushed into `AstDisplay for IntervalValue` in 6eddafce1d22c4476e07a7043a1a7312b05bf81b

ggevay commented on src/sql-parser/src/parser.rs:
```diff
@@ -3218,18 +3218,102 @@ impl<'a> Parser<'a> {
+    ) -> Result<WithOptionValue<Raw>, ParserError> {
+        let _ = self.consume_token(&Token::Eq);
+
+        if self.parse_keyword(ON) {
```
Fixed: 909ca77997ef7352491188b3636223e7eb0a8e38

ggevay commented on src/sql-parser/src/parser.rs:
```diff
@@ -3218,18 +3218,102 @@ impl<'a> Parser<'a> {
+                            "Invalid value for REFRESH EVERY: `{v}`. The value should be a \
+                            string parseable as an interval, e.g., '1 day'. The INTERVAL keyword \
+                            should NOT be present!"
+                        )
```
refactored away

ggevay commented on src/sql-parser/src/parser.rs:
```diff
@@ -3218,18 +3218,102 @@ impl<'a> Parser<'a> {
+                            "Invalid value for REFRESH EVERY: `{v}`. The value should be a \
+                            string parseable as an interval, e.g., '1 day'."
+                        )
+                    )
```
refactored away

ggevay commented on src/sql-parser/src/parser.rs:
```diff
@@ -3218,18 +3218,102 @@ impl<'a> Parser<'a> {
+                    self.peek_token().map(|token| token.to_string()).unwrap_or("".to_string()),
+                )
+            )
+        }
```
refactored away

ggevay commented on src/sql/src/pure.rs:
```diff
@@ -1475,3 +1482,239 @@ async fn compile_proto(
+
+/// Returns true if the [MaterializedViewOption] either already involves `mz_now()` or will involve
+/// after purification.
+pub fn materialized_view_option_contains_temporal(mvo: &MaterializedViewOption<Aug>) -> bool {
```
I'm not an expert, but I heard that calling the timestamp oracle is not free. (Especially with the cockroach-backed implementation maybe?) Would it be ok to call it even when the statement won't need it?

jkosh44 replied:
The timestamp oracle is expensive to call and we should try and avoid calling it unless we actually need a timestamp.

ggevay replied:
Let's consolidate this discussion [here](https://github.com/MaterializeInc/materialize/pull/23870#discussion_r1434228217).

benesch replied:
Ack, will mark this one as resolved.

ggevay commented on src/adapter/src/coord/command_handler.rs:
```diff
@@ -683,6 +688,87 @@ impl Coordinator {
+                    Ok(plan) => self.sequence_plan(ctx, plan, resolved_ids).await,
+                    Err(e) => ctx.retire(Err(e)),
+                }
+            }
```
I created a separate function for two reasons:

1. I didn't want to unconditionally pass the current time, to avoid the cost of querying the timestamp oracle. @jkosh44 [says](https://github.com/MaterializeInc/materialize/pull/23870#discussion_r1433012494) calling the timestamp oracle is costly. What do you think, @benesch? If we decide that we don't want to unconditionally call the timestamp oracle, then I don't see a way to avoid extending the SQL crate's public API: Even if I were to somehow merge my purification function into the already existing purification function, I'd have to first call `materialized_view_option_contains_temporal` to decide whether to call the timestamp oracle before the call of the merged purification function.

2. The existing purification function is async, but I didn't want to perform my purification async, because I was afraid of the chosen timestamp becoming invalid by the time the async function comes back. I'm not actually sure whether this could happen. I thought it could because `message_purified_statement_ready` has a comment "DROP CONNECTION may have sneaked in", which made me scared that other things may also sneak in that invalidate the timestamp. Is this fear valid or am I misunderstanding something?

benesch replied:
Ok, fair enough. Thinking about it a bit more, it's the inconsistency in the purification API that bothers me more than the expansion of the API surface area—i.e., that `purify_statement` _seems_ to be a thing that can handle all statements, but in fact it can only handle `CREATE SOURCE`, `ALTER SOURCE`, and `CREATE SINK`, and now there will be a totally separate path for `CREATE MATERIALIZED VIEW`.

An alternate way of slicing this would be to have four separate functions:

  * `async fn purify_create_source()`
  * `async fn purify_alter_source()`
  * `async fn purify_create_sink()`
  * `fn purify_create_materialized_view()`

Then it would be more clear that the adapter needs to handle each one of these functions slightly differently. As it stands, `message_purified_statement_ready` is pretty tangled up because handling the results of `purify_{create|alter}_source` has almost nothing in common with handling `purify_create_sink`.

I did a quick exploration of what would be required to do the above refactoring and ... it's grim. At least 1-2 days of work, and that's optimistic. So, let's put a pin in this entire line of thinking and forget I said anything.

----

> I was afraid of the chosen timestamp becoming invalid by the time the async function comes back. I'm not actually sure whether this could happen.

There can be an arbitrary delay, for sure, between when purification starts and when the coordinator handles the "purification ready" message. What it means for the chosen timestamp to become invalid here is your wheelhouse, though! Is the idea that selecting `mz_now()` as the oracle read timestamp is "safe" because the adapter always has read holds on every object at `mz_now()` and there is no gap between observing the oracle read timestamp and planning/installing the materialized view in the catalog (at which point it will install its own read holds for the next refresh)? If so, you're right to be worried about using `async` here, and having a sync function that never yields control of the coordinator's main loop is definitely the way to go.

ggevay replied:
> I did a quick exploration of what would be required to do the above refactoring and ... it's grim. At least 1-2 days of work, and that's optimistic. So, let's put a pin in this entire line of thinking and forget I said anything.

Ok. I've added a code comment linking your above message.

> There can be an arbitrary delay, for sure, between when purification starts and when the coordinator handles the "purification ready" message. What it means for the chosen timestamp to become invalid here is your wheelhouse, though! Is the idea that selecting mz_now() as the oracle read timestamp is "safe" because the adapter always has read holds on every object at mz_now() and there is no gap between observing the oracle read timestamp and planning/installing the materialized view in the catalog (at which point it will install its own read holds for the next refresh)?

Exactly.

> If so, you're right to be worried about using async here, and having a sync function that never yields control of the coordinator's main loop is definitely the way to go.

Ok, I've extended the doc comment of `purify_create_materialized_view_options` with the above reasoning of why it should not be async.

This is the commit ("Comments around purification"): https://github.com/MaterializeInc/materialize/pull/23870/commits/e6c152c7221bc483ad845e16ddb27c247f554f7a

ggevay commented on test/sqllogictest/materialized_views.slt:
```diff
@@ -645,7 +645,123 @@ SELECT * FROM mv_assertion_at_begin ORDER BY x;
+query error db error: ERROR: window functions are not allowed in REFRESH EVERY \.\.\. ALIGNED TO \(function pg_catalog\.lag\)
+CREATE MATERIALIZED VIEW mv_bad WITH (REFRESH EVERY '1 day' ALIGNED TO lag(7) OVER ()) AS SELECT 1;
+
+query error Error parsing interval of REFRESH EVERY: Unsupported value: RParen
```
refactored away

ggevay commented on src/sql/src/plan/with_options.rs:
```diff
@@ -426,27 +426,31 @@ impl<V: TryFromValue<Value>, T: AstInfo + std::fmt::Debug> TryFromValue<WithOpti
+            | WithOptionValue::Refresh(_) => sql_bail!(
                 "incompatible value types: cannot convert {} to {}",
                 match v {
+                    WithOptionValue::Value(_) => unreachable!(),
+                    WithOptionValue::Ident(_) => unreachable!(),
+                    WithOptionValue::RetainHistoryFor(_) => unreachable!(),
```
done

jkosh44 commented on src/sql/src/plan/with_options.rs:
```diff
@@ -510,3 +514,23 @@ impl ImpliedValue for Vec<KafkaBroker<Aug>> {
     }
 }
+
+impl TryFromValue<WithOptionValue<Aug>> for RefreshOptionValue<Aug> {
+    fn try_from_value(v: WithOptionValue<Aug>) -> Result<Self, PlanError> {
+        if let WithOptionValue::Refresh(r) = v {
+            Ok(r)
+        } else {
+            sql_bail!("cannot use value `{}` for a refresh option", v)
+        }
+    }
+
+    fn name() -> String {
+        "refresh option value".to_string()
+    }
+}
+
+impl ImpliedValue for RefreshOptionValue<Aug> {
+    fn implied_value() -> Result<Self, PlanError> {
+        sql_bail!("must provide a refresh option value")
+    }
+}
```
Yeah I think that's fine.

jkosh44 commented:
I've completely lost track of the details for this PR. Is there any specific questions/things I'm supposed to look at?

ggevay commented on src/adapter/src/catalog/migrate.rs:
```diff
@@ -95,6 +98,10 @@ pub(crate) async fn migrate(
 
+            if catalog_version <= Version::new(0, 80, u64::MAX) {
+                ast_rewrite_create_materialized_view_refresh_options_0_80_0(stmt)?;
+            }
```
Yes, I've updated it.

ggevay commented on src/adapter/src/coord/command_handler.rs:
```diff
@@ -693,6 +698,85 @@ impl Coordinator {
+                    // overestimation here by grabbing the ids of all indexes that are on ids
+                    // involved in the query. (We'd have to recursively follow view definitions,
+                    // similarly to `validate_timeline_context`.)
+                    // 2. For a peak, when the `least_valid_read` is later than the oracle's
+                    // `read_ts`, then the peak doesn't return before it completes at the chosen
```
fixed

ggevay commented on src/sql/src/catalog.rs:
```diff
@@ -196,7 +196,9 @@ pub trait SessionCatalog: fmt::Debug + ExprHumanizer + Send + Sync + ConnectionR
 
-    /// Resolves a partially-specified item name.
+    /// Resolves a partially-specified item name, that is NOT a function or
+    /// type. (For resolving functions or types, please use resolve_function
+    /// or resolve_type.)
```
fixed

ggevay commented on src/sql-parser/src/ast/defs/value.rs:
```diff
@@ -92,49 +92,56 @@ impl AstDisplay for Value {
 impl_display!(Value);
 
+impl AstDisplay for IntervalValue {
+    fn fmt<W: fmt::Write>(&self, f: &mut AstFormatter<W>) {
```
Done: https://github.com/MaterializeInc/materialize/pull/23870/commits/6eddafce1d22c4476e07a7043a1a7312b05bf81b

ggevay commented on src/sql-parser/src/ast/defs/statement.rs:
```diff
@@ -3186,11 +3188,66 @@ impl<T: AstInfo> AstDisplay for WithOptionValue<T> {
+            }) => {
+                f.write_str("EVERY '");
+                if f.redacted() {
+                    f.write_str("<REDACTED>'");
```
Done: https://github.com/MaterializeInc/materialize/pull/23870/commits/6eddafce1d22c4476e07a7043a1a7312b05bf81b

ggevay commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -2070,9 +2074,129 @@ pub fn plan_create_materialized_view(
+                    interval,
+                    aligned_to,
+                }) => {
+                    let interval = strconv::parse_interval(&interval.value)?; ///// todo: call Duration::try_from_value instead
```
Thanks a lot!

Done in 5daa7197a537d65c23dfcc2f55d297c5f7f78729 and 67f93859f1cd8d85ec74c3343479230a4d313225

ggevay commented on src/adapter/src/coord/command_handler.rs:
```diff
@@ -683,6 +688,87 @@ impl Coordinator {
+                    // past).
+                } else {
+                    None
+                };
```
In the meantime, the
```
                    let timeline = match timeline_context {
                        TimelineContext::TimelineDependent(timeline) => timeline,
                        TimelineContext::TimestampDependent
                        | TimelineContext::TimestampIndependent => {
                            // We default to EpochMilliseconds, similarly to `determine_timestamp_for`.
                            // Note that we didn't accurately decide whether we are TimestampDependent
                            // or TimestampIndependent, because for this we'd need to also check whether
                            // `query.contains_temporal()`, similarly to how `peek_stage_validate` does.
                            // However, this doesn't matter here, as we are just going to default to
                            // EpochMilliseconds in both cases.
                            Timeline::EpochMilliseconds
                        }
                    };
```
part got simplified into
```
                    // We default to EpochMilliseconds, similarly to `determine_timestamp_for`,
                    // but even in the TimestampIndependent case.
                    // Note that we didn't accurately decide whether we are TimestampDependent
                    // or TimestampIndependent, because for this we'd need to also check whether
                    // `query.contains_temporal()`, similarly to how `peek_stage_validate` does.
                    // However, this doesn't matter here, as we are just going to default to
                    // EpochMilliseconds in both cases.
                    let timeline = timeline_context
                        .timeline()
                        .unwrap_or(&Timeline::EpochMilliseconds);
```

If we were to allow this code to not support anything else then `EpochMilliseconds`, then we could further simplify by deleting even the
```
                    let timeline_context =
                        match self.validate_timeline_context(resolved_ids.0.clone()) {
                            Ok(tc) => tc,
                            Err(e) => return ctx.retire(Err(e)),
                        };
```
part, and then simply using `Timeline::EpochMilliseconds` in the `get_timestamp_oracle` call. But since this is already written, maybe it could just stay as it is, and then it would be a bit more future proof than simply assuming `EpochMilliseconds`?

benesch replied:
Ack, works for me.

ggevay commented on src/repr/src/lib.rs:
```diff
@@ -104,6 +104,7 @@ pub mod explain;
 pub mod fixed_length;
 pub mod global_id;
 pub mod namespaces;
+pub mod refresh_schedule;
```
Ok, I've moved it over to `expr`.

ggevay commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -2086,9 +2090,125 @@ pub fn plan_create_materialized_view(
+                        sql_bail!(
+                            "REFRESH interval too big: {}. Currently, only intervals not larger than 27 days are supported.",
+                            interval
+                        );
```
Refactored away

ggevay commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -2086,9 +2090,125 @@ pub fn plan_create_materialized_view(
+                        .into_literal_mz_timestamp().
+                        ok_or_else(|| sql_err!(
+                            "REFRESH AT argument must be an expression that can be simplified and/or cast \
+                            to a constant whose type is mz_timestamp (calling mz_now() is allowed)"
```
Ok, I've changed it into a structured error.

ggevay commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -2086,9 +2090,125 @@ pub fn plan_create_materialized_view(
+                        .ok_or_else(||
+                            sql_err!(
+                                "REFRESH EVERY ... ALIGNED TO argument must be an expression that can be simplified and/or cast \
+                                to a constant whose type is mz_timestamp (calling mz_now() is allowed)"
```
Ok, I've changed it into a structured error.

ggevay commented:
This is ready for a hopefully final round of review.

@benesch, could you please check https://github.com/MaterializeInc/materialize/pull/23870#discussion_r1439674786 ?

@jkosh44, I think I addressed all your comments.

Thanks a lot for the reviews!

jkosh44 commented on src/expr/src/refresh_schedule.rs:
```diff
@@ -0,0 +1,35 @@
+
+use mz_repr::Timestamp;
+use serde::{Deserialize, Serialize};
+use std::time::Duration;
```
We usually separate the `std` imports into their own block.

ggevay replied:
Yes, sorry, fixed!

benesch replied:
Super minor, but I think this is still outstanding!

ggevay replied:
Sorry, I resolved it locally, just haven't pushed it yet, because I was waiting for a Nightly.

benesch commented on src/sql-parser/src/ast/defs/statement.rs:
```diff
@@ -3186,11 +3188,62 @@ impl<T: AstInfo> AstDisplay for WithOptionValue<T> {
+            }
+        }
+    }
+}
```
This is very clean.

benesch commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -2079,9 +2084,125 @@ pub fn plan_create_materialized_view(
+                    interval,
+                    aligned_to,
+                }) => {
+                    let interval = Interval::try_from_value(Value::Interval(interval))?;
+                    if interval.as_microseconds() <= 0 {
+                        sql_bail!("REFRESH interval must be positive; got: {}", interval);
+                    }
+                    if interval.months != 0 {
+                        // This limitation is because we want Intervals to be cleanly convertable
+                        // to a unix epoch timestamp difference. When the interval involves months, then
+                        // this is not true anymore, because months have variable lengths.
+                        // See `Timestamp::round_up`.
+                        sql_bail!("REFRESH interval must not involve units larger than days");
+                    }
```
This cleaned up exactly as I envisioned it would with `Interval::try_from_value`! Nice!

benesch commented on src/sql/src/plan/statement/ddl.rs:
```diff
@@ -2079,9 +2084,125 @@ pub fn plan_create_materialized_view(
+        // Note: Seeing no REFRESH options at all (not even REFRESH ON COMMIT) should be acceptable:
+        // even though purification inserts REFRESH ON COMMIT if no other REFRESH option was given,
+        // we can't rely on this behavior in planning, because this won't happen for old
+        // materialized views that were created before this feature was introduced.
```
Don't block on this, but .. won't the migration ensure that even old matviews get `REFRESH ON COMMIT` before they ever hit planning?

ggevay replied:
Ah yes, thx, I forgot to delete this comment when I added the migration. I'll delete it before merging the PR.

philip-stoev reviewed (dismissed):
All the newly-introduced syntax needs to be in a Platform Checks test, so that its preservation across restarts and upgrades can be validated.

ggevay commented:
@philip-stoev, I'm planning to add more tests later (e.g. with [the follow-up pull request](https://github.com/MaterializeInc/materialize/pull/23819) that handles the Compute part of the new syntax). I'll make sure to include Platform Checks! Note that the feature is behind a feature flag, which will definitely be disabled for now, because this current PR implements only the Adapter part of the feature.

I'd like to merge this current PR as soon as possible, because I've been getting an insane amount of merge conflicts.
